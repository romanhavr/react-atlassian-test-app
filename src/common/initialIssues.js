export const initialIssues = [
  {
    id: 1,
    content: {
      id: 1,
      issue: '1-st',
      priority: { level: 2, label: 'Medium', value: 'medium' },
      assignee: 1,
      labelIds: [1]
    }
  },
  {
    id: 2,
    content: {
      id: 2,
      issue: '2-nd',
      priority: { level: 3, label: 'Low', value: 'low' },
      assignee: 2,
      labelIds: [2]
    }
  },
  {
    id: 3,
    content: {
      id: 3,
      issue: '1-st copy',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 1,
      labelIds: [1]
    }
  },
  {
    id: 4,
    content: {
      id: 4,
      issue: '3-rd',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 4,
      labelIds: [2, 3]
    }
  },
  {
    id: 5,
    content: {
      id: 5,
      issue: '1-st another copy',
      priority: { level: 3, label: 'Low', value: 'low' },
      assignee: 3,
      labelIds: [1, 3]
    }
  },
  {
    id: 6,
    content: {
      id: 6,
      issue: '1-st',
      priority: { level: 1, label: 'High', value: 'high' },
      assignee: 4,
      labelIds: [1, 3]
    }
  },
  {
    id: 7,
    content: {
      id: 7,
      issue: '1-st',
      priority: { level: 2, label: 'Medium', value: 'medium' },
      assignee: 2,
      labelIds: [1, 3]
    }
  }
]