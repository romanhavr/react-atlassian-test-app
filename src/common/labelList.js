export const labels = [
    { id: 1, name: 'label1', label: 'Label 1' },
    { id: 2, name: 'anotherLabel', label: 'Another Label' },
    { id: 3, name: 'newLabel', label: 'New Label' },
]