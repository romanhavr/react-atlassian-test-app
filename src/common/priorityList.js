export const priorityList = [
    { level: 1, label: 'High', value: 'high' },
    { level: 2, label: 'Medium', value: 'medium' },
    { level: 3, label: 'Low', value: 'low' }
]