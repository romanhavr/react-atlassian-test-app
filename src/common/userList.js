// @flow

import type { User } from '../interfaces/interfaces';

export const userList: User[] = [
    {
        id: 1,
        displayName: 'Admin',
        avatar: 'https://pickaface.net/gallery/avatar/unr_edwinavator_170412_0152_yoidv.png'
    },
    {
        id: 2,
        displayName: 'Elvis',
        avatar: 'https://pickaface.net/gallery/avatar/Nirvana53183d8f601d7.png'
    },
    {
        id: 3,
        displayName: 'Roling Stone',
        avatar: 'https://tradiesinbusiness.com.au/wp-content/uploads/2019/09/avatar-jongen-voorbeeld-1.jpg'
    },
    {
        id: 4,
        displayName: 'John Smith',
        avatar: 'https://apollodesign.nl/wp-content/uploads/2015/08/avatar-casper-tekening.png'
    },
]