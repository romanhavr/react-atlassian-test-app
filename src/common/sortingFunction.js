// @flow

import type { TableItem, SortingOptions, SortingCriteria } from '../interfaces/interfaces';
import { priorityList } from '../common/priorityList';
import { userList } from '../common/userList';

export function sortingFunc(a: TableItem, b: TableItem, sortingBy: SortingOptions, i: number): number {
        if (comparing(a, b, sortingBy[i].value, 'more')) {
            return 1
        } else if (comparing(a, b, sortingBy[i].value, 'less')) {
            return -1
        } else if (comparing(a, b, sortingBy[i].value, 'eq') && i < sortingBy.length - 1) {
            return sortingFunc(a, b, sortingBy, ++i)
        } else
            return 0
};

function comparing (a: TableItem, b: TableItem, sortingValue: string, compareIndex: string): boolean {
    const newObject: SortingCriteria = sortingCriteria (a.content[sortingValue], b.content[sortingValue], sortingValue)
    switch (compareIndex) {
        case 'more' :
            return newObject.newObj1 > newObject.newObj2;
        case 'eq' :
            return newObject.newObj1 === newObject.newObj2;
        case 'less' :
            return newObject.newObj1 < newObject.newObj2;
        default:
            return newObject.newObj1 > newObject.newObj2;
    }
}

function sortingCriteria (obj1, obj2, sortingValue: string): SortingCriteria {
    switch(sortingValue) {
        case 'issue' :
            return { newObj1: obj1, newObj2: obj2};
        case 'assignee' :
            const userById1 = userList.find(el => el.id === obj1);
            const userById2 = userList.find(el => el.id === obj2);
            if (userById1 && userById2) return { 
                newObj1: userById1.displayName,
                newObj2: userById2.displayName
            };
            break;
        case 'priority' :
            const priorityByValue1 = priorityList.find(el => el.value === obj1.value);
            const priorityByValue2 = priorityList.find(el => el.value === obj2.value);
            if (priorityByValue1 && priorityByValue2) return {
                newObj1: priorityByValue1.level,
                newObj2: priorityByValue2.level
            };
            break;
        default:
            return { newObj1: obj1, newObj2: obj2};
    }
}