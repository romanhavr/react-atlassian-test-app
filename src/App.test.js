import { shallow, mount } from 'enzyme';
import React from 'react';
import ConnectedApp, { App } from './App';
import { initialIssues } from './common/initialIssues';
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

const mockStore = configureMockStore();

const chosenItem = {
    id: 1572444816584,
    issue: "1-st copy",
    priority: 1,
    assignee: 1,
    labelIds: [1]
}

const initialDispatches = {
    sortTable: jest.fn(),
    userFetchReq: jest.fn()
}

const initialSagaStore = {
    sagaData: null,
    sagaSucceeded: null,
    sagaFailed: null,
    someUser: null,
}

const store = mockStore({
    issues: initialIssues,
    ui: { chosenItem },
    saga: initialSagaStore,
    ...initialDispatches
});

const wrapper = mount(
    <Provider store={store}>
        <App {...initialDispatches} />
    </Provider>
);

describe('App component testing...', () => {

    const creatIssueButton = wrapper.find('button[data-test="create-issue-button"]');
    const sagaActionButton = wrapper.find('button[data-test="saga-action-button"]');

    it('App should "click" Create Issue Button', () => {
        expect(creatIssueButton.length).toBe(1);

        creatIssueButton.simulate('click');
        expect(initialDispatches.sortTable).toHaveBeenLastCalledWith(null)
    })

    it('App should "click" Saga Action Button', () => {
        expect(sagaActionButton.length).toBe(1);

        sagaActionButton.simulate('click', 'SAGA Action');
        expect(initialDispatches.userFetchReq).toHaveBeenLastCalledWith('SAGA Action')
    })

    it('App should include SNAPSHOT', () => {
        const tree = shallow(<header></header>);
        expect(tree).toMatchSnapshot();
    })
});

let connectedWrapper, connectedStore;

describe('App Redux testing', () => {

    beforeEach( () => {
        connectedStore = store;
        connectedWrapper = mount(
            <Provider store={connectedStore}>
                <ConnectedApp />
            </Provider>
        );
    })
    
    it('App props should match with initialState', () => {
        expect(connectedWrapper.prop('store').getState().saga).toBe(initialSagaStore)
    });

    
    it('App actions should dispatch', () => {
        console.log(connectedStore.dispatch)
        let action;
        connectedStore.dispatch(sortTable({
            label: 'Assignee',
            value: 'assignee'
        }));
        connectedStore.dispatch(userFetchReq('SAGA Action'));
        action = connectedStore.getActions();
        expect(action[0].type).toBe("SORT_TABLE");
        expect(action[1].type).toBe("USER_FETCH_REQUESTED");
    })
})