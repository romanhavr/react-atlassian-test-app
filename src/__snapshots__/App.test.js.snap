// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component testing... App should include SNAPSHOT 1`] = `<header />`;
